# Fixing data

At this step we will clear the data that may disturb us later

In [261]:
import pandas as pd
import scipy as sc
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt      
import math
import seaborn as sns
from collections import Counter


At this stage he gave the dataframe in terms of basic problems so that we can continue to work optimally

In [262]:
df = pd.read_csv("players_Attributes.csv")
df

column_1     value              name                           club  \
0            0  €180.00m     Kylian Mbappé            Paris Saint-Germain   
1            1  €170.00m    Erling Haaland                Manchester City   
2            2  €120.00m   Vinicius Junior                    Real Madrid   
3            3   €90.00m        Harry Kane              Tottenham Hotspur   
4            4   €85.00m       Rafael Leão                       AC Milan   
...        ...       ...               ...                            ...   
1377      1377    €5.00m    Federico Gatti                    Juventus FC   
1378      1378    €5.00m     Adrià Pedrosa         RCD Espanyol Barcelona   
1379      1379    €5.00m  Joaquín Piquerez  Sociedade Esportiva Palmeiras   
1380      1380    €5.00m     Nicolas Cozza                Montpellier HSC   
1381      1381    €5.00m       Marco Burch                      FC Luzern   

           nation        position  age  \
0          France  Centre-Forward   24   
1          Norway  Centre-Forward   22   
2          Brazil     Left Winger   22   
3         England  Centre-Forward   29   
4        Portugal     Left Winger   23   
...           ...             ...  ...   
1377        Italy     Centre-Back   24   
1378        Spain       Left-Back   24   
1379      Uruguay       Left-Back   24   
1380       France       Left-Back   23   
1381  Switzerland     Centre-Back   22   

                                                   link  Corners  Crossing  \
0     https://fminside.net/players/3-fm-23/85139014-...       13        13   
1     https://fminside.net/players/3-fm-23/29179241-...        7        10   
2     https://fminside.net/players/3-fm-23/19302146-...        7        13   
3     https://fminside.net/players/3-fm-23/28049320-...        9        15   
4     https://fminside.net/players/3-fm-23/83111483-...        5        15   
...                                                 ...      ...       ...   
1377  https://fminside.net/players/3-fm-23/43491384-...        7         7   
1378  https://fminside.net/players/3-fm-23/95035991-...        7         9   
1379  https://fminside.net/players/3-fm-23/78077014-...       11        12   
1380  https://fminside.net/players/3-fm-23/49037761-...        9        11   
1381  https://fminside.net/players/3-fm-23/84156311-...        6         7   

      ...  Balance  Jumping Reach  Natural Fitness  Pace  Stamina  Strength  \
0     ...       15              8               15    20       14        11   
1     ...       17             18               19    19       14        17   
2     ...       14              8               15    17       14        12   
3     ...       14             13               16    14       16        15   
4     ...       16             14               15    16       11        14   
...   ...      ...            ...              ...   ...      ...       ...   
1377  ...       12             15               14    13       14        15   
1378  ...       13              8               10    15       13        10   
1379  ...       14             12               13    13       13        14   
1380  ...       14             12               13    12       14        13   
1381  ...       13             12               13    13       14        12   

      rating   Foot  Length  Weight  
0         94  Right     178      73  
1         93   Left     195      88  
2         87  Right     177      62  
3         92   Both     188      86  
4         82  Right     188      81  
...      ...    ...     ...     ...  
1377      65  Right     190      84  
1378      66   Left     176      69  
1379      66   Left     183      81  
1380      63   Left     178      72  
1381      59   Left     183      75  

[1382 rows x 48 columns]

In [263]:
df.drop(columns="column_1",inplace=True)

In [264]:
df2=df.copy()
df2.drop(columns='link',inplace=True)

We will remove the link column that we removed with Selenium because it is no longer relevant for us

The price column contains unnecessary symbols that may interfere later (we will work with values in the millions)

In [265]:
#Removing unnesecery symbols
#df2['value']=df2['value'].replace('m','').replace('€','')
#df2['value']=df2['value'].astype('float')
df2['value'] = df2['value'].replace(r'\€', '', regex= True)
df2['value'] = df2['value'].replace('m', '', regex= True)
df2

value              name                           club       nation  \
0     180.00     Kylian Mbappé            Paris Saint-Germain       France   
1     170.00    Erling Haaland                Manchester City       Norway   
2     120.00   Vinicius Junior                    Real Madrid       Brazil   
3      90.00        Harry Kane              Tottenham Hotspur      England   
4      85.00       Rafael Leão                       AC Milan     Portugal   
...      ...               ...                            ...          ...   
1377    5.00    Federico Gatti                    Juventus FC        Italy   
1378    5.00     Adrià Pedrosa         RCD Espanyol Barcelona        Spain   
1379    5.00  Joaquín Piquerez  Sociedade Esportiva Palmeiras      Uruguay   
1380    5.00     Nicolas Cozza                Montpellier HSC       France   
1381    5.00       Marco Burch                      FC Luzern  Switzerland   

            position  age  Corners  Crossing  Dribbling  Finishing  ...  \
0     Centre-Forward   24       13        13         18         17  ...   
1     Centre-Forward   22        7        10         14         18  ...   
2        Left Winger   22        7        13         18         16  ...   
3     Centre-Forward   29        9        15         14         19  ...   
4        Left Winger   23        5        15         17         15  ...   
...              ...  ...      ...       ...        ...        ...  ...   
1377     Centre-Back   24        7         7          8         10  ...   
1378       Left-Back   24        7         9         14          6  ...   
1379       Left-Back   24       11        12         12         10  ...   
1380       Left-Back   23        9        11         11          8  ...   
1381     Centre-Back   22        6         7          8          7  ...   

      Balance  Jumping Reach  Natural Fitness  Pace  Stamina  Strength  \
0          15              8               15    20       14        11   
1          17             18               19    19       14        17   
2          14              8               15    17       14        12   
3          14             13               16    14       16        15   
4          16             14               15    16       11        14   
...       ...            ...              ...   ...      ...       ...   
1377       12             15               14    13       14        15   
1378       13              8               10    15       13        10   
1379       14             12               13    13       13        14   
1380       14             12               13    12       14        13   
1381       13             12               13    13       14        12   

      rating   Foot  Length  Weight  
0         94  Right     178      73  
1         93   Left     195      88  
2         87  Right     177      62  
3         92   Both     188      86  
4         82  Right     188      81  
...      ...    ...     ...     ...  
1377      65  Right     190      84  
1378      66   Left     176      69  
1379      66   Left     183      81  
1380      63   Left     178      72  
1381      59   Left     183      75  

[1382 rows x 46 columns]

A little information about the dataframe we discussed

In [266]:
print(df2.info())
arr_teams=df2['club'].unique()
arr_nations=df2['nation'].unique()
print(len(arr_teams),len(arr_nations))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1382 entries, 0 to 1381
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   value             1382 non-null   object
 1   name              1382 non-null   object
 2   club              1382 non-null   object
 3   nation            1382 non-null   object
 4   position          1382 non-null   object
 5   age               1382 non-null   int64 
 6   Corners           1382 non-null   int64 
 7   Crossing          1382 non-null   int64 
 8   Dribbling         1382 non-null   int64 
 9   Finishing         1382 non-null   int64 
 10  First Touch       1382 non-null   int64 
 11  Free Kick Taking  1382 non-null   int64 
 12  Heading           1382 non-null   int64 
 13  Long Shots        1382 non-null   int64 
 14  Long Throws       1382 non-null   int64 
 15  Marking           1382 non-null   int64 
 16  Passing           1382 non-null   int64 
 17  Penalty Taking

The value is object and not number

In [267]:
df2['value']=df2['value'].astype(float)


In [268]:
df2['Weight']=df2['Weight'].astype(float)

In [269]:
print(df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1382 entries, 0 to 1381
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   value             1382 non-null   float64
 1   name              1382 non-null   object 
 2   club              1382 non-null   object 
 3   nation            1382 non-null   object 
 4   position          1382 non-null   object 
 5   age               1382 non-null   int64  
 6   Corners           1382 non-null   int64  
 7   Crossing          1382 non-null   int64  
 8   Dribbling         1382 non-null   int64  
 9   Finishing         1382 non-null   int64  
 10  First Touch       1382 non-null   int64  
 11  Free Kick Taking  1382 non-null   int64  
 12  Heading           1382 non-null   int64  
 13  Long Shots        1382 non-null   int64  
 14  Long Throws       1382 non-null   int64  
 15  Marking           1382 non-null   int64  
 16  Passing           1382 non-null   int64  


In [270]:
df2['value'].sort_values().head()

1381    5.0
1359    5.0
1360    5.0
1361    5.0
1362    5.0
Name: value, dtype: float64

In [271]:
df2.sort_values(by='value',ascending=False).head()

value             name                 club   nation          position  \
0    180.0    Kylian Mbappé  Paris Saint-Germain   France    Centre-Forward   
1    170.0   Erling Haaland      Manchester City   Norway    Centre-Forward   
2    120.0  Vinicius Junior          Real Madrid   Brazil       Left Winger   
462  110.0  Jude Bellingham    Borussia Dortmund  England  Central Midfield   
463  110.0       Phil Foden      Manchester City  England  Central Midfield   

     age  Corners  Crossing  Dribbling  Finishing  ...  Balance  \
0     24       13        13         18         17  ...       15   
1     22        7        10         14         18  ...       17   
2     22        7        13         18         16  ...       14   
462   19       10        11         16         13  ...       14   
463   22       13        13         17         14  ...       15   

     Jumping Reach  Natural Fitness  Pace  Stamina  Strength  rating   Foot  \
0                8               15    20       14        11      94  Right   
1               18               19    19       14        17      93   Left   
2                8               15    17       14        12      87  Right   
462             12               16    15       16        11      78  Right   
463              7               14    16       15         9      82   Left   

     Length  Weight  
0       178    73.0  
1       195    88.0  
2       177    62.0  
462     186    75.0  
463     171    64.0  

[5 rows x 46 columns]

Scanning thק data

In [272]:
df2.iloc[:,0:25].describe()

value          age      Corners     Crossing    Dribbling  \
count  1382.000000  1382.000000  1382.000000  1382.000000  1382.000000   
mean     18.239725    24.918958     8.895080    10.851664    12.636758   
std      17.346809     3.424253     3.535051     2.696122     2.403782   
min       5.000000    16.000000     1.000000     3.000000     3.000000   
25%       8.000000    22.000000     6.000000     9.000000    11.000000   
50%      12.000000    25.000000     9.000000    11.000000    13.000000   
75%      20.000000    27.000000    12.000000    13.000000    14.000000   
max     180.000000    37.000000    20.000000    18.000000    20.000000   

         Finishing  First Touch  Free Kick Taking      Heading   Long Shots  \
count  1382.000000  1382.000000       1382.000000  1382.000000  1382.000000   
mean     10.672214    13.227207          9.360347    10.774240    10.722865   
std       3.026615     1.765929          3.368782     3.296979     2.857572   
min       3.000000     8.000000          1.000000     2.000000     2.000000   
25%       8.000000    12.000000          7.000000     8.000000     9.000000   
50%      11.000000    13.000000          9.000000    11.000000    11.000000   
75%      13.000000    14.000000         12.000000    13.000000    13.000000   
max      19.000000    20.000000         20.000000    20.000000    19.000000   

       ...      Marking      Passing  Penalty Taking     Tackling  \
count  ...  1382.000000  1382.000000     1382.000000  1382.000000   
mean   ...     9.719247    13.138205       10.337916    10.770622   
std    ...     3.676950     1.729039        3.213746     3.690437   
min    ...     1.000000     7.000000        1.000000     2.000000   
25%    ...     6.000000    12.000000        8.000000     7.000000   
50%    ...    10.000000    13.000000       10.000000    12.000000   
75%    ...    13.000000    14.000000       12.000000    14.000000   
max    ...    18.000000    20.000000       20.000000    20.000000   

         Technique   Aggression  Anticipation      Bravery    Composure  \
count  1382.000000  1382.000000   1382.000000  1382.000000  1382.000000   
mean     13.598408    12.185962     13.210564    12.901592    12.963097   
std       1.827483     2.847971      1.724662     2.486257     1.754506   
min       6.000000     4.000000      6.000000     4.000000     5.000000   
25%      12.000000    10.000000     12.000000    11.250000    12.000000   
50%      14.000000    12.000000     13.000000    13.000000    13.000000   
75%      15.000000    14.000000     14.000000    15.000000    14.000000   
max      20.000000    20.000000     20.000000    20.000000    19.000000   

       Concentration  
count    1382.000000  
mean       12.240232  
std         1.824732  
min         5.000000  
25%        11.000000  
50%        12.000000  
75%        13.000000  
max        18.000000  

[8 rows x 21 columns]

In [273]:
df2.iloc[:,25:].describe()

Decisions  Determination        Flair   Leadership  Off the Ball  \
count  1382.000000    1382.000000  1382.000000  1382.000000   1382.000000   
mean     12.489870      13.932706    12.177279     9.855282     12.657742   
std       1.711096       2.373328     3.232211     3.048849      2.572780   
min       4.000000       2.000000     2.000000     2.000000      3.000000   
25%      12.000000      13.000000    10.000000     8.000000     12.000000   
50%      13.000000      14.000000    13.000000    10.000000     13.000000   
75%      13.000000      15.000000    14.000000    12.000000     14.000000   
max      18.000000      20.000000    20.000000    19.000000     20.000000   

       Positioning     Teamwork       Vision    Work Rate  Acceleration  \
count  1382.000000  1382.000000  1382.000000  1382.000000   1382.000000   
mean     10.696093    13.461650    12.473227    13.794501     13.699711   
std       3.239908     2.139499     2.172254     2.092993      1.786663   
min       2.000000     5.000000     3.000000     6.000000      7.000000   
25%       8.000000    12.000000    11.000000    13.000000     13.000000   
50%      11.000000    14.000000    12.000000    14.000000     14.000000   
75%      13.000000    15.000000    14.000000    15.000000     15.000000   
max      18.000000    20.000000    20.000000    20.000000     20.000000   

           Agility      Balance  Jumping Reach  Natural Fitness         Pace  \
count  1382.000000  1382.000000    1382.000000      1382.000000  1382.000000   
mean     13.251085    13.125904      11.078871        13.914616    13.814038   
std       1.752597     1.907357       3.179223         1.914147     1.644161   
min       7.000000     5.000000       2.000000         5.000000     7.000000   
25%      12.000000    12.000000       9.000000        13.000000    13.000000   
50%      13.000000    13.000000      11.000000        14.000000    14.000000   
75%      14.000000    14.000000      14.000000        15.000000    15.000000   
max      19.000000    20.000000      20.000000        20.000000    20.000000   

           Stamina     Strength       rating       Length       Weight  
count  1382.000000  1382.000000  1382.000000  1382.000000  1382.000000  
mean     13.876266    12.139653    68.797395   181.420405    71.569465  
std       1.669647     2.710246     6.426281     6.690031    16.948336  
min       7.000000     2.000000    25.000000   160.000000     0.000000  
25%      13.000000    10.000000    65.000000   176.000000    70.000000  
50%      14.000000    12.000000    68.000000   182.000000    74.000000  
75%      15.000000    14.000000    72.000000   186.000000    79.000000  
max      20.000000    20.000000    94.000000   204.000000   100.000000

The minimum value of the weight is 0 which does not make sense and we would like to deal with it

In [274]:
df2.loc[df2['Weight'] == 0]

value                name                 club       nation  \
47     40.0      Anthony Gordon           Everton FC      England   
84     25.0       Nico Williams      Athletic Bilbao        Spain   
119    20.0     Brennan Johnson    Nottingham Forest        Wales   
148    18.0       Eddie Nketiah           Arsenal FC      England   
159    17.0  Keane Lewis-Potter         Brentford FC      England   
...     ...                 ...                  ...          ...   
1355    5.5      Armando Obispo        PSV Eindhoven  Netherlands   
1364    5.0     Ahmetcan Kaplan       Ajax Amsterdam       Turkey   
1366    5.0       Calvin Ramsay         Liverpool FC     Scotland   
1369    5.0   Charlie Cresswell          Millwall FC      England   
1375    5.0    Timothée Pembélé  Paris Saint-Germain       France   

            position  age  Corners  Crossing  Dribbling  Finishing  ...  \
47       Left Winger   21       11        13         14         11  ...   
84      Right Winger   20        7        13         14         13  ...   
119   Second Striker   21        8        13         14         13  ...   
148   Centre-Forward   23        2         7         14         14  ...   
159      Left Winger   21       11        13         14         14  ...   
...              ...  ...      ...       ...        ...        ...  ...   
1355     Centre-Back   23        8        11          9          5  ...   
1364     Centre-Back   19        4         6          9          6  ...   
1366      Right-Back   19       13        14         14          8  ...   
1369     Centre-Back   20        2         5         11          8  ...   
1375      Right-Back   20        6        13         13         10  ...   

      Balance  Jumping Reach  Natural Fitness  Pace  Stamina  Strength  \
47          7             11               10    14       14         9   
84         11             10               14    14       13        14   
119        12              8               14    15       13         9   
148        13              9               14    17       14        10   
159        13             10               14    14       13         9   
...       ...            ...              ...   ...      ...       ...   
1355       13             13               14    12       16        13   
1364       14             14               13    11       12        13   
1366       12             10               11    14       10         9   
1369       13             15               13    13       13        16   
1375       14             14               15    14       13        13   

      rating   Foot  Length  Weight  
47        66  Right     183     0.0  
84        70   Left     180     0.0  
119       68  Right     180     0.0  
148       69  Right     175     0.0  
159       64  Right     170     0.0  
...      ...    ...     ...     ...  
1355      64   Left     186     0.0  
1364      55   Left     189     0.0  
1366      55  Right     179     0.0  
1369      59  Right     190     0.0  
1375      60  Right     182     0.0  

[62 rows x 46 columns]

In [275]:
df2.loc[df2['Weight'] != 0,['Length','Weight']].describe()

Length       Weight
count  1320.000000  1320.000000
mean    181.406818    74.931061
std       6.688370     6.976334
min     160.000000    56.000000
25%     176.000000    70.000000
50%     182.000000    75.000000
75%     186.000000    80.000000
max     204.000000   100.000000

We have problem with the Weight because we have some rows that the value is 0.
We saw a proportion between height and weight 
The average length is 181 and the average weight is 75 approx
181/75=2.4

This way of treatment seems more logical to us than using the median or average .
Due to the fact that soccer players are usually in good physical shape and there is a direct relationship between height and weight.

In [276]:
df2.loc[:,['Length','Weight']].describe()

Length       Weight
count  1382.000000  1382.000000
mean    181.420405    71.569465
std       6.690031    16.948336
min     160.000000     0.000000
25%     176.000000    70.000000
50%     182.000000    74.000000
75%     186.000000    79.000000
max     204.000000   100.000000

In [277]:
df2.loc[df2['Weight'] == 0,['Weight']] = df2.loc[df2['Weight'] == 0,['Length']]/2.4
df2.loc[:,['Length','Weight']].describe()

Length       Weight
count  1382.000000  1320.000000
mean    181.420405    74.931061
std       6.690031     6.976334
min     160.000000    56.000000
25%     176.000000    70.000000
50%     182.000000    75.000000
75%     186.000000    80.000000
max     204.000000   100.000000

# We will check if there are any missing values

In [278]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1382 entries, 0 to 1381
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   value             1382 non-null   float64
 1   name              1382 non-null   object 
 2   club              1382 non-null   object 
 3   nation            1382 non-null   object 
 4   position          1382 non-null   object 
 5   age               1382 non-null   int64  
 6   Corners           1382 non-null   int64  
 7   Crossing          1382 non-null   int64  
 8   Dribbling         1382 non-null   int64  
 9   Finishing         1382 non-null   int64  
 10  First Touch       1382 non-null   int64  
 11  Free Kick Taking  1382 non-null   int64  
 12  Heading           1382 non-null   int64  
 13  Long Shots        1382 non-null   int64  
 14  Long Throws       1382 non-null   int64  
 15  Marking           1382 non-null   int64  
 16  Passing           1382 non-null   int64  


In [279]:
df2['Length'].isnull().unique()

array([False])

In [280]:
df2['Weight'].isnull().unique()

array([False,  True])

In [281]:
df2[df2['Weight'].isnull()]

value                name                 club       nation  \
47     40.0      Anthony Gordon           Everton FC      England   
84     25.0       Nico Williams      Athletic Bilbao        Spain   
119    20.0     Brennan Johnson    Nottingham Forest        Wales   
148    18.0       Eddie Nketiah           Arsenal FC      England   
159    17.0  Keane Lewis-Potter         Brentford FC      England   
...     ...                 ...                  ...          ...   
1355    5.5      Armando Obispo        PSV Eindhoven  Netherlands   
1364    5.0     Ahmetcan Kaplan       Ajax Amsterdam       Turkey   
1366    5.0       Calvin Ramsay         Liverpool FC     Scotland   
1369    5.0   Charlie Cresswell          Millwall FC      England   
1375    5.0    Timothée Pembélé  Paris Saint-Germain       France   

            position  age  Corners  Crossing  Dribbling  Finishing  ...  \
47       Left Winger   21       11        13         14         11  ...   
84      Right Winger   20        7        13         14         13  ...   
119   Second Striker   21        8        13         14         13  ...   
148   Centre-Forward   23        2         7         14         14  ...   
159      Left Winger   21       11        13         14         14  ...   
...              ...  ...      ...       ...        ...        ...  ...   
1355     Centre-Back   23        8        11          9          5  ...   
1364     Centre-Back   19        4         6          9          6  ...   
1366      Right-Back   19       13        14         14          8  ...   
1369     Centre-Back   20        2         5         11          8  ...   
1375      Right-Back   20        6        13         13         10  ...   

      Balance  Jumping Reach  Natural Fitness  Pace  Stamina  Strength  \
47          7             11               10    14       14         9   
84         11             10               14    14       13        14   
119        12              8               14    15       13         9   
148        13              9               14    17       14        10   
159        13             10               14    14       13         9   
...       ...            ...              ...   ...      ...       ...   
1355       13             13               14    12       16        13   
1364       14             14               13    11       12        13   
1366       12             10               11    14       10         9   
1369       13             15               13    13       13        16   
1375       14             14               15    14       13        13   

      rating   Foot  Length  Weight  
47        66  Right     183     NaN  
84        70   Left     180     NaN  
119       68  Right     180     NaN  
148       69  Right     175     NaN  
159       64  Right     170     NaN  
...      ...    ...     ...     ...  
1355      64   Left     186     NaN  
1364      55   Left     189     NaN  
1366      55  Right     179     NaN  
1369      59  Right     190     NaN  
1375      60  Right     182     NaN  

[62 rows x 46 columns]

In [282]:
df2[df2['Weight'].isnull()]

value                name                 club       nation  \
47     40.0      Anthony Gordon           Everton FC      England   
84     25.0       Nico Williams      Athletic Bilbao        Spain   
119    20.0     Brennan Johnson    Nottingham Forest        Wales   
148    18.0       Eddie Nketiah           Arsenal FC      England   
159    17.0  Keane Lewis-Potter         Brentford FC      England   
...     ...                 ...                  ...          ...   
1355    5.5      Armando Obispo        PSV Eindhoven  Netherlands   
1364    5.0     Ahmetcan Kaplan       Ajax Amsterdam       Turkey   
1366    5.0       Calvin Ramsay         Liverpool FC     Scotland   
1369    5.0   Charlie Cresswell          Millwall FC      England   
1375    5.0    Timothée Pembélé  Paris Saint-Germain       France   

            position  age  Corners  Crossing  Dribbling  Finishing  ...  \
47       Left Winger   21       11        13         14         11  ...   
84      Right Winger   20        7        13         14         13  ...   
119   Second Striker   21        8        13         14         13  ...   
148   Centre-Forward   23        2         7         14         14  ...   
159      Left Winger   21       11        13         14         14  ...   
...              ...  ...      ...       ...        ...        ...  ...   
1355     Centre-Back   23        8        11          9          5  ...   
1364     Centre-Back   19        4         6          9          6  ...   
1366      Right-Back   19       13        14         14          8  ...   
1369     Centre-Back   20        2         5         11          8  ...   
1375      Right-Back   20        6        13         13         10  ...   

      Balance  Jumping Reach  Natural Fitness  Pace  Stamina  Strength  \
47          7             11               10    14       14         9   
84         11             10               14    14       13        14   
119        12              8               14    15       13         9   
148        13              9               14    17       14        10   
159        13             10               14    14       13         9   
...       ...            ...              ...   ...      ...       ...   
1355       13             13               14    12       16        13   
1364       14             14               13    11       12        13   
1366       12             10               11    14       10         9   
1369       13             15               13    13       13        16   
1375       14             14               15    14       13        13   

      rating   Foot  Length  Weight  
47        66  Right     183     NaN  
84        70   Left     180     NaN  
119       68  Right     180     NaN  
148       69  Right     175     NaN  
159       64  Right     170     NaN  
...      ...    ...     ...     ...  
1355      64   Left     186     NaN  
1364      55   Left     189     NaN  
1366      55  Right     179     NaN  
1369      59  Right     190     NaN  
1375      60  Right     182     NaN  

[62 rows x 46 columns]

In [283]:
print(df2.loc[df2['Weight'].isnull(),'Weight'])

47     NaN
84     NaN
119    NaN
148    NaN
159    NaN
        ..
1355   NaN
1364   NaN
1366   NaN
1369   NaN
1375   NaN
Name: Weight, Length: 62, dtype: float64


In [284]:
#df2.loc[df2[df2['Weight'].isnull()],['Weight']] = df2.loc[df2[df2['Weight'].isnull()],['Length']]/2.4
#df2['Weight'].fillna(df.Weight.median(),inplace=True)
df2.loc[df2['Weight'].isnull(),'Weight']=df2.loc[df2['Weight'].isnull(),'Length']/2.4

df2['Weight'].isnull().unique()

array([False])

In [285]:
for col in df2.columns:
    if(len(df2[col].isnull().unique())>1):
        print(col)

We have no missing values

In [286]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1382 entries, 0 to 1381
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   value             1382 non-null   float64
 1   name              1382 non-null   object 
 2   club              1382 non-null   object 
 3   nation            1382 non-null   object 
 4   position          1382 non-null   object 
 5   age               1382 non-null   int64  
 6   Corners           1382 non-null   int64  
 7   Crossing          1382 non-null   int64  
 8   Dribbling         1382 non-null   int64  
 9   Finishing         1382 non-null   int64  
 10  First Touch       1382 non-null   int64  
 11  Free Kick Taking  1382 non-null   int64  
 12  Heading           1382 non-null   int64  
 13  Long Shots        1382 non-null   int64  
 14  Long Throws       1382 non-null   int64  
 15  Marking           1382 non-null   int64  
 16  Passing           1382 non-null   int64  


In [287]:
df2.to_csv("players_clean_1.csv",index=False)